src/accounts.rs

```diff
@@ -228,30 +237,60 @@ impl Accounts {
 
     /// Unified event emitter.
     pub async fn get_event_emitter(&self) -> EventEmitter {
-        let emitters: Vec<_> = self
-            .accounts
-            .read()
-            .await
-            .iter()
-            .map(|(_id, a)| a.get_event_emitter())
-            .collect();
-
-        EventEmitter(futures::stream::select_all(emitters))
+        self.emitter.clone()
     }
 }
 
-#[derive(Debug)]
-pub struct EventEmitter(futures::stream::SelectAll<crate::events::EventEmitter>);
+#[derive(Debug, Clone)]
+pub struct EventEmitter {
+    /// Aggregate stream of events from all accounts.
+    stream: Arc<RwLock<futures::stream::SelectAll<crate::events::EventEmitter>>>,
+
+    /// Sender for the channel where new account emitters will be pushed.
+    sender: Sender<crate::events::EventEmitter>,
+
+    /// Receiver for the channel where new account emitters will be pushed.
+    receiver: Receiver<crate::events::EventEmitter>,
+}
 
 impl EventEmitter {
+    pub fn new() -> Self {
+        let (sender, receiver) = async_std::channel::unbounded();
+        Self {
+            stream: Arc::new(RwLock::new(futures::stream::SelectAll::new())),
+            sender,
+            receiver,
+        }
+    }
+
     /// Blocking recv of an event. Return `None` if all `Sender`s have been droped.
     pub fn recv_sync(&mut self) -> Option<Event> {
-        async_std::task::block_on(self.recv())
+        async_std::task::block_on(self.recv()).unwrap_or_default()
+    }
+
+    /// Async recv of an event. Return `None` if all `Sender`s have been dropped.
+    pub async fn recv(&mut self) -> Result<Option<Event>> {
+        let mut stream = self.stream.write().await;
+        loop {
+            match futures::future::select(self.receiver.recv(), stream.next()).await {
```

Comment:
you will loose events here, because this will pick up from both sources, and so the one that finishes later, will get dropped

see https://docs.rs/futures/0.3.16/futures/future/fn.select.html for more details

Reply:
using `race` from async-std should work, it drops the other future, so stops trying to receive on the non slected part

Reply:
Both `race` and `select` look like they are doing the same to me, except that to use `race` I will have to wrap values into `Either` manually, while `select` is doing it for me. Select runs until completion of one of the futures, and returns a pair of value and the second unfinished future. I drop the second future (it gets into `_` pattern). `race` just always drops the second future.

The question is if `recv` and `next` are atomic, i.e. is it guaranteed that if I drop them before they return that they have not extracted the next value from the channel/stream yet. As long as they don't return control to executor between extracting the value and returning it, both solutions (`select` and `race`) should be ok.